File: fb/240.Search-a-2D-Matrix-II.cpp
// Time O(m+n)
// Space O(1)
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int row = matrix.size() - 1;
        int col = 0;

        while (row >= 0 && col < matrix[0].size()) {
            if (matrix[row][col] > target) {
                row--;
            } else if (matrix[row][col] < target) {
                col++;
            } else { // found it
                return true;
            }
        }

        return false;
    }
};

File: fb/240.Search-a-2D-Matrix-II.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        # start from the left bottom corner
        row = len(matrix) - 1
        col = 0
        
        while row >= 0 and col < len(matrix[0]):
            if matrix[row][col] > target:
                row -= 1
            elif matrix[row][col] < target:
                col += 1
            else: 
                return True
            
        return False
